File: day3.c
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

int main()
{
    FILE* file = fopen("day3.txt", "r");
    int sum = 0, sum2 = 0;
    bool enable = true;
    char line[4096];

    while (fgets(line, sizeof(line), file)) {
        for (char* s = line; *s; s++) {
            int x, y, n = 0;
            if (2 == sscanf(s, "mul(%d,%d)%n", &x, &y, &n) && n) {
                sum += x * y;
                if (enable) sum2 += x * y;
            } else if (!strncmp(s, "do()", 4)) {
                enable = true;
            } else if (!strncmp(s, "don't()", 7)) {
                enable = false;
            }
        }
    }

    printf("%d %d\n", sum, sum2);
}

File: day3.lua
#!/usr/bin/env lua

local file = io.open("day3.txt")
local sum, sum2 = 0, 0
local enable = true

for line in file:lines() do
    for i = 1, #line do
        local x, y = string.match(line, "^mul%((%d+),(%d+)%)", i)
        if x and y then
            local prod = tonumber(x) * tonumber(y)
            sum = sum + prod
            if enable then sum2 = sum2 + prod end
        elseif string.match(line, "^do%(%)", i) then
            enable = true
        elseif string.match(line, "^don't%(%)", i) then
            enable = false
        end
    end
end

print(sum, sum2)
